helper/exptime/backoff.go

```diff
@@ -0,0 +1,142 @@
+// Package exptime provides a generalized exponential backoff retry implementation.
+//
+// This package was copied from oss.indeed.com/go/libtime/decay and modified.
+package exptime
+
+import (
+	"errors"
+	"fmt"
+	"math/rand"
+	"time"
+)
+
+var (
+	// ErrMaximumTimeExceeded indicates the maximum wait time has been exceeded.
+	ErrMaximumTimeExceeded = errors.New("maximum backoff time exceeded")
+)
+
+// A TryFunc is what gets executed between retry wait periods during execution
+// of Backoff. The keepRetrying return value is used to control whether a retry
+// attempt should be made. This feature is useful in manipulating control flow
+// in cases where it is known a retry will not be successful.
+type TryFunc func() (keepRetrying bool, err error)
+
+// BackoffOptions allow for fine-tuning backoff behavior.
+type BackoffOptions struct {
+	// MaxSleepTime represents the maximum amount of time
+	// the exponential backoff system will spend sleeping,
+	// accumulating the amount of time spent asleep between
+	// retries.
+	//
+	// The algorithm starts at an interval of InitialGapSize
+	// and increases exponentially (x2 each iteration) from there.
+	// With no jitter, a MaxSleepTime of 10 seconds and InitialGapSize
+	// of 1 millisecond would suggest a total of 15 attempts
+	// (since the very last retry truncates the sleep time to
+	// align exactly with MaxSleepTime).
+	MaxSleepTime time.Duration
+
+	// InitialGapSize sets the initial amount of time the algorithm
+	// will sleep before the first retry (after the first attempt).
+	// The actual amount of sleep time will include a random amount
+	// of jitter, if MaxJitterSize is non-zero.
+	InitialGapSize time.Duration
+
+	// MaxJitterSize limits how much randomness we may
+	// introduce in the duration of each retry interval.
+	// The purpose of introducing jitter is to mitigate the
+	// effect of thundering herds
+	MaxJitterSize time.Duration
+
+	// RandomSeed is used for generating a randomly computed
+	// jitter size for each retry.
+	RandomSeed int64
+
+	// Sleeper is used to cause the process to sleep for
+	// a computed amount of time. If not set, a default
+	// implementation based on time.Sleep will be used.
+	Sleeper Sleeper
+}
+
+// A Sleeper is a useful way for calling time.Sleep
+// in a mock-able way for tests.
+type Sleeper func(time.Duration)
```

Comment:
It looks like this alias is the only diff between `oss.indeed.com/go/libtime/decay` and this source file? While I understand and agree with not wanting to pull in the whole darn `libtime` library for this, the "derived work" being created here is an awfully minimal change. So for licensing safety, I think we're going to need to include the upstream [license file](https://github.com/indeedeng/libtime/blob/master/LICENSE) in this package.

Reply:
:+1: added the license

But yeah, the only reason I don't want to pull it in as a module is the transitive dependency on github.com/gojuno/minimock/v3 is several orders of magnitude larger than this one function we actually want